### ISSUE TITLE
Segmentation Fault/Win 7: Happens when writing to Shapefile

### ISSUE THREAD

**Issue by leowmjw** [Toblerity/Fiona#303]
Windows 7, 32-bit, Python 2.7
GDAL + Fiona from http://www.lfd.uci.edu/~gohlke/pythonlibs/#fiona

The code was working fine on OSX and is being ported over to Windows.
The structure of the data map is shown further below:
```
        sink = fiona.open(
            ModifiedShapefile.shape_filename + '.shp', 'w',
            crs=from_epsg(3857),
            driver=ModifiedShapefile.sink_driver,
            schema=ModifiedShapefile.sink_schema,
        )
...
        for norm_dm in sorted(new_feature_map, key=ModifiedShapefile._keyify):
...
                    sink.write(new_feature_map[norm_dm])
...
...
        sink.close()

```

At first I wondered if it is some sort of bad data but same problem is encountered even if this particular row is skipped; so it must be something else (memory exhaustion?).  Full dataset is small (< 900 rows).

The data item right before the SegFault:
```
DM is 222/81/01
{'geometry': {'coordinates': [[(115.40043648415751, 4.174299525846915),
                               (115.40043648415751, 4.174299525846915),
                               (115.41136975191978, 4.185203486118631),
                               (115.41136975191978, 4.185203486118631),
                               (115.42277837848597, 4.1970554366188),
                               (115.42277837848597, 4.1970554366188),
                               (115.42515517609898, 4.205114594781534),
                               (115.42515517609898, 4.205114594781534),
                               (115.430384130129, 4.20938122119581),
                               (115.430384130129, 4.20938122119581),
                               (115.44084203818896, 4.202270389005245),
                               (115.44084203818896, 4.202270389005245),
                               (115.45367674386196, 4.1942113782357),
                               (115.45367674386196, 4.1942113782357),
                               (115.46413465102361, 4.204166922409789),
                               (115.46413465102361, 4.204166922409789),
                               (115.47649399699436, 4.209855821385591),
                               (115.47649399699436, 4.209855821385591),
                               (115.49027942117362, 4.213174403315818),
                               (115.49027942117362, 4.213174403315818),
                               (115.49788517281661, 4.222181561611432),
                               (115.49788517281661, 4.222181561611432),
                               (115.50786772117429, 4.228344349581833),
                               (115.50786772117429, 4.228344349581833),
                               (115.51927634863887, 4.225974229775585),
                               (115.51927634863887, 4.225974229775585),
                               (115.53543856953276, 4.235455383637585),
                               (115.53543856953276, 4.235455383637585),
                               (115.54494575908652, 4.24493634181017),
                               (115.54494575908652, 4.24493634181017),
                               (115.55635438565271, 4.244936475046226),
                               (115.55635438565271, 4.244936475046226),
                               (115.57299196714725, 4.232137498340967),
                               (115.57299196714725, 4.232137498340967),
                               (115.58630203162417, 4.23450788619178),
                               (115.58630203162417, 4.23450788619178),
                               (115.58867882833887, 4.247781084172304),
                               (115.58867882833887, 4.247781084172304),
                               (115.5962845799819, 4.253943636786112),
                               (115.5962845799819, 4.253943636786112),
                               (115.61244680177413, 4.26200235708098),
                               (115.61244680177413, 4.26200235708098),
                               (115.61624967669734, 4.281437331350181),
                               (115.61624967669734, 4.281437331350181),
                               (115.61862647431036, 4.289969608902236),
                               (115.61862647431036, 4.289969608902236),
                               (115.62670758475731, 4.29707983819052),
                               (115.62670758475731, 4.29707983819052),
                               (115.63193653878729, 4.305611972386565),
                               (115.63193653878729, 4.305611972386565),
                               (115.63193653878729, 4.319357886310071),
                               (115.63193653878729, 4.319357886310071),
                               (115.63621477431104, 4.333577581881658),
                               (115.63621477431104, 4.333577581881658),
                               (115.63954229043031, 4.351114809480884),
                               (115.63954229043031, 4.351114809480884),
                               (115.66426098237174, 4.333103925054179),
                               (115.66426098237174, 4.333103925054179),
                               (115.66853921699722, 4.317936334053929),
                               (115.66853921699722, 4.317936334053929),
                               (115.66901457669947, 4.303716397279724),
                               (115.66901457669947, 4.303716397279724),
                               (115.66758849849104, 4.280489896957755),
                               (115.66758849849104, 4.280489896957755),
                               (115.65522915252029, 4.259158678477431),
                               (115.65522915252029, 4.259158678477431),
                               (115.64572196296652, 4.243989439785524),
                               (115.64572196296652, 4.243989439785524),
                               (115.64524660416258, 4.232612389066733),
                               (115.64524660416258, 4.232612389066733),
                               (115.6461973226688, 4.199428387144438),
                               (115.6461973226688, 4.199428387144438),
                               (115.63811621222183, 4.180939435215351),
                               (115.63811621222183, 4.180939435215351),
                               (115.63906693072808, 4.16766512258453),
                               (115.63906693072808, 4.16766512258453),
                               (115.65427843311579, 4.151072077187174),
                               (115.65427843311579, 4.151072077187174),
                               (115.64952483878807, 4.133530228094622),
                               (115.64952483878807, 4.133530228094622),
                               (115.63764085251961, 4.121677303924022),
                               (115.63764085251961, 4.121677303924022),
                               (115.62860902266812, 4.116461916434083),
                               (115.62860902266812, 4.116461916434083),
                               (115.61529895819115, 4.108875833325837),
                               (115.61529895819115, 4.108875833325837),
                               (115.61007000416113, 4.098919132300094),
                               (115.61007000416113, 4.098919132300094),
                               (115.60816856625043, 4.078531312161702),
                               (115.60816856625043, 4.078531312161702),
                               (115.60198889371422, 4.06762592314153),
                               (115.60198889371422, 4.06762592314153),
                               (115.58725275013042, 4.065729159357095),
                               (115.58725275013042, 4.065729159357095),
                               (115.57964699938573, 4.069996396558145),
                               (115.57964699938573, 4.069996396558145),
                               (115.57013980983199, 4.064306512124192),
                               (115.57013980983199, 4.064306512124192),
                               (115.56110797998048, 4.066203002508477),
                               (115.56110797998048, 4.066203002508477),
                               (115.55825582356353, 4.075685878081416),
                               (115.55825582356353, 4.075685878081416),
                               (115.54922399371203, 4.086590978289946),
                               (115.54922399371203, 4.086590978289946),
                               (115.54874863490805, 4.099866672752112),
                               (115.54874863490805, 4.099866672752112),
                               (115.5444703993843, 4.111719738229108),
                               (115.5444703993843, 4.111719738229108),
                               (115.53638928893731, 4.119305542616137),
                               (115.53638928893731, 4.119305542616137),
                               (115.52165314535357, 4.137795693130399),
                               (115.52165314535357, 4.137795693130399),
                               (115.5188009889366, 4.150596396708892),
                               (115.5188009889366, 4.150596396708892),
                               (115.51737491072811, 4.172878641978779),
                               (115.51737491072811, 4.172878641978779),
                               (115.5078677211743, 4.179515676348664),
                               (115.5078677211743, 4.179515676348664),
                               (115.49503301639965, 4.185204463258118),
                               (115.49503301639965, 4.185204463258118),
                               (115.49123014057814, 4.165292956858812),
                               (115.49123014057814, 4.165292956858812),
                               (115.49503301639965, 4.150596119100232),
                               (115.49503301639965, 4.150596119100232),
                               (115.48980406236966, 4.134950684908535),
                               (115.48980406236966, 4.134950684908535),
                               (115.49408229699513, 4.117882698048365),
                               (115.49408229699513, 4.117882698048365),
                               (115.48647654625044, 4.109348451646413),
                               (115.48647654625044, 4.109348451646413),
                               (115.47887079460742, 4.118830753664087),
                               (115.47887079460742, 4.118830753664087),
                               (115.47506791878591, 4.135898726589496),
                               (115.47506791878591, 4.135898726589496),
                               (115.47174040266664, 4.143010250154877),
                               (115.47174040266664, 4.143010250154877),
                               (115.4646100107259, 4.152966247905147),
                               (115.4646100107259, 4.152966247905147),
                               (115.45415210266592, 4.166240701160047),
                               (115.45415210266592, 4.166240701160047),
                               (115.43561308415897, 4.180937070080771),
                               (115.43561308415897, 4.180937070080771),
                               (115.4213523002775, 4.177144263511485),
                               (115.4213523002775, 4.177144263511485),
                               (115.40756687609826, 4.170981021384566),
                               (115.40756687609826, 4.170981021384566),
                               (115.40043648415751, 4.174299525846915),
                               (115.40043648415751, 4.174299525846915)]],
              'type': 'Polygon'},
 'id': '815',
 'properties': OrderedDict([(u'NAME', u'222/81/01'), (u'NAMA_DM', u'SEMADOH'), (u'PAR_LAMA', u'222'), (u'DUN_LAMA', u'70'), (u'DM_LAMA', u'02'), (u'PENGUNDI', 661), (u'KODPAR', u'222'), (u'PAR_BARU', u'LAWAS'), (u'KODDUN', u'81'), (u'DUN_BARU', u'Ba\u2019kelalan'), (u'KODDM', u'01'), (u'DM_BARU', u'SEMADOH'), (u'PENGUNDI_B', u'674'), ('NAMA_LAMA', u'222/70/02')]),
 'type': 'Feature'}

Process finished with exit code -1073741819 (0xC0000005)

```

Any ideas on how/where to put a breakpoint in FIona to figure out why the Segmentation Fault is happening?

**Comment by leowmjw** [comment]
Using the FaultHandler module; was able to identify the root cause as the writerecs function in fiona/collection.py:
```
    def writerecords(self, records):
        """Stages multiple records for writing to disk."""
        if self.closed:
            raise ValueError("I/O operation on closed collection")
        if self.mode not in ('a', 'w'):
            raise IOError("collection not open for writing")
        self.session.writerecs(records, self)
        self._len = self.session.get_length()
        self._bounds = self.session.get_extent()
```

The root cause is the Unicode character of the text "Ba’kelalan" in the field DUN_BARU as per below:
```
OrderedDict([(u'NAME', u'222/81/01'), (u'NAMA_DM', u'SEMADOH'), (u'PAR_LAMA', u'222'), (u'DUN_LAMA', u'70'), (u'DM_LAMA', u'02'), (u'PENGUNDI', 661), (u'KODPAR', u'222'), (u'PAR_BARU', u'LAWAS'), (u'KODDUN', u'81'), (u'DUN_BARU', u'Ba\u2019kelalan'), (u'KODDM', u'01'), (u'DM_BARU', u'SEMADOH'), (u'PENGUNDI_B', u'674'), ('NAMA_LAMA', u'222/70/02')])
```

Again; it only fails on Windows, works fine in OSX.

**Comment by sgillies** [comment]
Here's the root of the problem: you can't encode that string in ISO-8859-1, the default encoding for shapefiles:

```python
>>> u'Ba\u2019kelalan'.encode('ISO-8859-1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    UnicodeError: 'ascii' codec won't decode byte 0x00 in position -1.
UnicodeEncodeError: 'latin-1' codec can't encode character u'\u2019' in position 2: ordinal not in range(256)
```

And then Fiona doesn't do the right thing at https://github.com/Toblerity/Fiona/blob/3e817159cc8c14bf54c6df47896fa69f24566cff/fiona/ogrext1.pyx#L312.

Good news is there's a work around for you and that's to set the output encoding to UTF-8 when you open your writeable collection: `fiona.open('example.shp', 'w', ..., encoding='utf-8')`.

**Comment by sgillies** [comment]
Closed via #309
